core/config/preconditioner_config.cpp

```diff
@@ -0,0 +1,295 @@
+    std::string str("solver::LowerTrs");
+    if (auto& obj = config.get("l_solver_type")) {
+        str = obj.get_string();
+    }
+    if (str == "solver::LowerTrs") {
+        return dispatch<gko::LinOpFactory,
+                        IcHelper2<solver::LowerTrs>::Configurator>(
+            config, context, updated,
+            make_type_selector(updated.get_value_typestr(), value_type_list()),
+            make_type_selector(updated.get_index_typestr(), index_type_list()));
+    } else if (str == "solver::Ir") {
+        return dispatch<gko::LinOpFactory, IcHelper1<solver::Ir>::Configurator>(
+            config, context, updated,
+            make_type_selector(updated.get_value_typestr(), value_type_list()),
+            make_type_selector(updated.get_index_typestr(), index_type_list()));
+    } else if (str == "preconditioner::LowerIsai") {
+        return dispatch<gko::LinOpFactory,
+                        IcHelper2<preconditioner::LowerIsai>::Configurator>(
+            config, context, updated,
+            make_type_selector(updated.get_value_typestr(), value_type_list()),
+            make_type_selector(updated.get_index_typestr(), index_type_list()));
+    } else if (str == "solver::Gmres") {
+        return dispatch<gko::LinOpFactory,
+                        IcHelper1<solver::Gmres>::Configurator>(
+            config, context, updated,
+            make_type_selector(updated.get_value_typestr(), value_type_list()),
+            make_type_selector(updated.get_index_typestr(), index_type_list()));
+    } else {
+        GKO_INVALID_STATE("does not have valid LSolverType");
+    }
+}
+
+
+template <template <typename V> class LSolverBase,
+          template <typename V> class USolverBase, bool ReverseApply>
+class IluHelper1 {
+public:
+    template <typename ValueType, typename IndexType>
+    class Configurator
+        : public IluSolverHelper<
+              LSolverBase<ValueType>, USolverBase<ValueType>,
+              ReverseApply>::template Configurator<ValueType, IndexType> {};
+};
+
+
+template <template <typename V, typename I> class LSolverBase,
+          template <typename V, typename I> class USolverBase,
+          bool ReverseApply>
+class IluHelper2 {
+public:
+    template <typename ValueType, typename IndexType>
+    class Configurator
+        : public IluSolverHelper<
+              LSolverBase<ValueType, IndexType>,
+              USolverBase<ValueType, IndexType>,
+              ReverseApply>::template Configurator<ValueType, IndexType> {};
+};
+
+
+template <>
+deferred_factory_parameter<gko::LinOpFactory> parse<LinOpFactoryType::Ilu>(
+    const pnode& config, const registry& context, const type_descriptor& td)
+{
+    auto updated = update_type(config, td);
+    auto dispatch_solver = [&](auto reverse_apply)
+        -> deferred_factory_parameter<gko::LinOpFactory> {
+        using ReverseApply = decltype(reverse_apply);
+        // always use symmetric solver for USolverType
```

Comment:
is guess for now that is a sufficient simplification.

Reply:
I also prefer #1029 in the future without instantiate each type